libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,794 @@
+      y,
+      [&](ast::record& rec) -> failure_or<void> {
+        for (auto& i : rec.items) {
+          auto* field = try_as<ast::record::field>(i);
+          if (not field) {
+            diagnostic::error("cannot use `...` here").primary(y).emit(ctx);
+            return failure::promise();
+          }
+          const auto loc = field->expr.get_location();
+          TRY(match(
+            field->expr,
+            [&](ast::function_call& call) -> failure_or<void> {
+              args.y[field->name.name] = std::move(call);
+              return {};
+            },
+            [&](auto& expr) -> failure_or<void> {
+              if (args.res) {
+                diagnostic::error("an aggregation is required")
+                  .primary(field->expr)
+                  .emit(ctx);
+                return failure::promise();
+              }
+              auto result
+                = ast::function_call{entity, {std::move(expr)}, loc, false};
+              TENZIR_ASSERT(resolve_entities(result, ctx));
+              args.y[field->name.name] = std::move(result);
+              return {};
+            }));
+        }
+        return {};
+      },
+      [&](ast::function_call& call) -> failure_or<void> {
+        args.y["y"] = std::move(call);
+        return {};
+      },
+      [&](auto&) -> failure_or<void> {
+        if (args.res) {
+          diagnostic::error("an aggregation is required").primary(y).emit(ctx);
+          return failure::promise();
+        }
+        const auto loc = y.get_location();
+        auto result = ast::function_call{entity, {std::move(y)}, loc, false};
+        TENZIR_ASSERT(resolve_entities(result, ctx));
+        args.y["y"] = std::move(result);
+        return {};
+      });
+  }
+
+  auto handle_limit(const chart_args& args, ast::binary_op op,
+                    const located<data>& limit, diagnostic_handler& dh) const
+    -> failure_or<xlimit> {
+    const auto& loc = limit.source;
+    auto result = match(
+      limit.inner,
+      [&](const caf::none_t&) -> failure_or<ast::constant> {
+        diagnostic::error("limit cannot be `null`").primary(limit).emit(dh);
+        return failure::promise();
+      },
+      [&](const pattern&) -> failure_or<ast::constant> {
+        diagnostic::error("limit cannot be a pattern").primary(limit).emit(dh);
+        return failure::promise();
+      },
+      [&](const duration& d) -> failure_or<ast::constant> {
+        if (args.res) {
+          const auto val = d.count();
+          const auto count = std::abs(args.res->inner.count());
+          const auto rem = std::abs(val % count);
+          if (rem) {
+            const auto ceil = val >= 0 ? count - rem : rem;
+            const auto floor = val >= 0 ? -rem : rem - count;
+            return ast::constant{
+              duration{val + (op == ast::binary_op::leq ? floor : ceil)},
+              loc,
+            };
+          }
+        }
+        return ast::constant{d, loc};
+      },
```

Comment:
IIRC, we wanted `x_min` to be floored, and `x_max` to be ceiled, right?

<img width="1223" alt="image" src="https://github.com/user-attachments/assets/6418e283-0664-47fe-b234-617a5fd302a6" />

Reply:
Right, it should have been `geq`. Fixed.